### PR TITLE
added merging methods and scoring functions

### PR DESCRIPTION
**Pull request by SaminYeasar**
* Added merging methods: 
  * SparseMerge
  * SLERP
  * LERP
  * TiesMerge: made correction
  * TaskArithmatic
  * ModelBreadcrumbs
  * UniformMerge

* Added different scoring functions
  * grow and drop
  * layer drop + sparse
  * model wise sparse
  * gradient-magnitude based sparse
  * weight-magnitude based sparse
  * added backwardhook: will mask gradient during backdrop

### CODE REVIEW

**Comment by SaminYeasar** [comment]
@microsoft-github-policy-service agree company="Microsoft"

File: mttl/models/modifiers/sparse_mask.py [hunk 1]
```diff
@@ -15,7 +15,63 @@
...
+    """
+    The hook function can't be defined directly here because of Python's
+    late binding which would result in all hooks getting the very last
+    mask! Getting it through another function forces early binding.
+    """
+    def hook(grads):
+        return grads * keep_mask
+    return hook
+
+def load_mask(f_name):
```
Comment (fverac) [inline mttl/models/modifiers/sparse_mask.py:30]: Is there a way we can avoid needing these functions `save_mask` and `load_mask`?

weight_mask should be included in state_dict by now so can be reloaded from checkpoints?

File: mttl/models/modifiers/sparse_mask.py [hunk 2]
```diff
@@ -116,67 +172,180 @@ def get_regular_sparse_mask(m):
...
+            if isinstance(m, SparseMaskModule):
+                keep_masks = (m.sparse_layer.weight_mask.grad>=accepted_score).float()
+                if  (keep_masks.sum()/m.sparse_layer.weight_mask.numel())>= m.keep_ratio:
+                    keep_masks = get_regular_sparse_mask(m)
+                    m.revert_weight_grad_and_update_mask(keep_masks)
+                else:
+                    keep_masks = torch.zeros_like(m.sparse_layer.weight)
+                    m.revert_weight_grad_and_update_mask(keep_masks)
+    # based on gradient-magnitude
+    elif parameter_selection_procedure=='gradient_magnitude':
```
Comment (fverac) [inline mttl/models/modifiers/sparse_mask.py:329]: can you add tests for the new permutations of these parameters in `test_sparse_mask.py`?

File: mttl/models/modifiers/sparse_mask.py [hunk 3]
```diff
@@ -241,6 +410,11 @@ def __init__(
     def patch_forward(self):
         self.sparse_layer.forward = types.MethodType(mod_forward, self.sparse_layer)
 
+    @torch.no_grad()
+    def convert_sparse_weight_to_1D(self):
```
Comment (fverac) [inline mttl/models/modifiers/sparse_mask.py:414]: Is this ever used? ctrl+f doesn't find anything for me.

File: mttl/models/modifiers/sparse_mask.py [hunk 4]
```diff
@@ -116,67 +172,180 @@ def get_regular_sparse_mask(m):
...
-    if parameter_selection_procedure == "per_layer":
-        for m in module.modules():
+    if parameter_selection_procedure=='per_layer':
+        for idx, m in enumerate(module.modules()):
             if isinstance(m, SparseMaskModule):
-                if m.sparse_cat == "block_sparse":
+                if m.sparse_cat == 'block_sparse':
                     keep_masks = get_block_mask(m)
-                elif m.sparse_cat == "regular_sparse":
+                    # check: sample noise-block-idx
```
Comment (fverac) [inline mttl/models/modifiers/sparse_mask.py:271]: remove comments when ready. optionally add logging

**Comment by SaminYeasar** [comment]
will make a new pr
